Order form

graffad opened the pull request:
ready to check

graffad committed:
order-form-layout

-basic layout order-fom
-added redux order config

graffad committed:
order-form-validation

- added masked fields
- added validation schemas
- enabled requests for : cities,countries,send cart
- added step 4 : result request

ValadzkoAliaksei commented on src/components/cart/Cart.js:
```diff
@@ -1,41 +1,155 @@
+    }
+  };
+
+  function closeCart(el) {
```
определись какие функции используешь стрелочные и натационные)

ValadzkoAliaksei commented on src/components/cart/Cart.js:
```diff
@@ -1,41 +1,155 @@
+
+  function closeCart(el) {
+    if (
+      el.classList.contains("cart--active") ||
```
нельзя так отображаться к элементу и его стилям. Ищи другие зависимости от которых можно выполнять необходимые действия

ValadzkoAliaksei commented on src/components/cart/Cart.js:
```diff
@@ -1,41 +1,155 @@
+
+  // close cart and reset
+  useEffect(() => {
+    document.addEventListener("click", (event) => {
```
нельзя к документу обращаться и вешать не него обработчики событий.
Например можно повесить маску и смотреть клик на ней или взять элемент в useRef и смотреть на клики вне этого элемента.

ValadzkoAliaksei commented on src/components/cart/Cart.js:
```diff
@@ -1,41 +1,155 @@
+  // reset or unregister step2
+  useEffect(() => {
+    setValue("country", "");
+    if (deliveryMethod === "pickup from post offices") {
```
явное сравнение одной и той же переменной со строкой... может тут удобней switch case?

ValadzkoAliaksei commented on src/components/cart/Cart.js:
```diff
@@ -1,41 +1,155 @@
   return (
-    <div className={`cart-outer ${isShowCart ? "cart--active" : ""}`} data-test-id="cart">
+    <div
+      className={`cart-outer ${isShowCart ? "cart--active" : ""}`}
```
classNames использовать

ValadzkoAliaksei commented on src/components/cart/Step2.js:
```diff
@@ -1,12 +1,411 @@
+                  }`}
+                >
+                  <Controller
+                    name="postcode"
```
почему то у инпута нет поля name на выходе, так же как и для поля phone

ValadzkoAliaksei commented on src/components/cart/Step2.js:
```diff
@@ -1,12 +1,411 @@
+                    render={({ field: { onChange, onBlur, value } }) => (
+                      <InputMask
+                        onChange={onChange}
+                        mask="aa 999999"
```
BY по умолчанию. пользователь может вводить только цифры 
так настояла дизайнер, и не очень то хотелось спорить.

ValadzkoAliaksei commented on src/components/cart/Step2.js:
```diff
@@ -1,12 +1,411 @@
+                      />
+                    )}
+                  />
+                  {errors?.postcode && (
```
тут и для phone ошибки появляются сразу, а должны по снятию фокуса или нажатию кнопки отправить. Так же как и все остальные поля.
А то сейчас все нормально работают, а эти нет.

ValadzkoAliaksei commented on src/redux/reducers/orderSlice.js:
```diff
@@ -0,0 +1,54 @@
+    // actions
+    reduxGetOrderStoresInfo(state, action) {
+      // payload : {type: countries or cities, search: {city:qwe, country:"sss"} }
+      state.isLoading = action.payload.type;
```
можно деструктуризировать на входе action что бы не писать такие длинные цепочки.

ValadzkoAliaksei commented on src/redux/reducers/orderSlice.js:
```diff
@@ -0,0 +1,54 @@
+      state.errorType = false;
+    },
+    reduxSetOrderStoresInfo(state, action) {
+      const resultCountries = action.payload.data.map((item) => ({
```
логика преобразования должна быть за редюсером.

ValadzkoAliaksei commented on src/redux/rootSaga.js:
```diff
@@ -2,11 +2,13 @@ import { all, fork } from "redux-saga/effects";
-        fork(sagaWatcherReview),
-        fork(sagaWatcherSubscribe)
-    ]);
+  yield all([
+    fork(sagaWatcherProducts),
+    fork(sagaWatcherReview),
+    fork(sagaWatcherSubscribe),
+    fork(sagaWatcherOrder),
+  ]);
```
Тут fork-ать не обязательно. можно просто
 yield all([
   sagaWatcherProducts(),
    sagaWatcherReview(),
    sagaWatcherSubscribe(),
    sagaWatcherOrder(),
  ]);

ValadzkoAliaksei commented on src/redux/sagas/sagaOrder.js:
```diff
@@ -0,0 +1,34 @@
+  reduxSetOrderError,
+} from "../reducers/orderSlice";
+
+function* sagaWorkerOrder({payload}) {
```
странное решение сделать одну сагу на 2 разных процесса и просто if ом их разделять.
Редюсер по-моему так же почему то сделан. Я бы все-таки разделил

ValadzkoAliaksei commented on src/services/masksWithConditions.js:
```diff
@@ -0,0 +1,103 @@
+  };
+  return (
+    <InputMask
+      mask={mask}
```
вот тут наверно name и нужен. поэтому тесты и не проходят.
Ну и запуск валидации так же

ValadzkoAliaksei commented on src/services/masksWithConditions.js:
```diff
@@ -0,0 +1,103 @@
+    />
+  );
+}
+
+
+
+
+
+
+
+
+
+
+
```
откуда столько пробелов?)

graffad committed:
order-form-fix-v1

-fixed input names
-fixed postcode mask
-fixed request cities
-form validation method is changed - onBlur
-fixed redux orderSlice + saga (split req)
-added classnames lib
-close cart behavior is changed

graffad committed:
order-form-fix-v2

-fixed card field 16 symbols + validation
-fixed viewCard step in step3

graffad committed:
order-form-fix-v3

-fixed react-select placeholder
-fixed viewCard step4 with reset

graffad committed:
order-form-fix-v4

-fixed react-select placeholder

graffad committed:
order-form-fix-v5

-removed react-select
-added custom-select

graffad committed:
order-form-fix-v6

-test: clear error on focus
-test: clear errors on radio change

graffad committed:
order-form-fix-v7

-test: clear error on focus eventListener

ValadzkoAliaksei commented on src/App.scss:
```diff
@@ -15,12 +15,11 @@ $grayDark: #9c9c9c;
   overflow: hidden;
 }
-@mixin scrollBar($widthOrHeight : "width", $size:3px){
+@mixin scrollBar($widthOrHeight: "width", $size: 3px) {
```
чаще всего CSS разделяют по компонентам. Что бы не было такого большого монстрика)

ValadzkoAliaksei commented on src/components/cart/Cart.js:
```diff
@@ -1,97 +1,244 @@
+import UserService from "../../services/userService";
 
-export default function Cart({ isShowCart }) {
+export default function Cart({ isShowCart, setIsShowCart }) {
```
чего решили тут не использовать стрелочную функцию?)

ValadzkoAliaksei commented on src/components/cart/Cart.js:
```diff
@@ -1,97 +1,244 @@
+      cardDate: "",
+      card: "",
+      cardCVV: "",
+    },
```
вынесите в отдельную переменную

ValadzkoAliaksei commented on src/components/cart/Cart.js:
```diff
@@ -1,97 +1,244 @@
+  async function onSubmit(data) {
+    setIsOrderError(false);
+    setOrderErrorMessage("");
+    console.log(data)
```
консольки почистить

ValadzkoAliaksei commented on src/components/cart/Step2.js:
```diff
@@ -1,12 +1,363 @@
+            </p>
+          )}
+        </div>
+        {(deliveryMethod === "pickup from post offices" ||
```
тестовые переменные лучше вынести в константы в виде объекта и брать их значение по ключу объекта

ValadzkoAliaksei commented on src/components/cart/Step3.js:
```diff
@@ -0,0 +1,210 @@
+import mcPng from "../../images/credit-cards/mastercard_x42.png";
+import eye from "../../images/eyeOpen.png";
+import eyeClose from "../../images/eyeClose.png";
+import { MaskCardDate } from "../../services/masksWithConditions";
```
импорт над фотками, я ж вроде писал про очередность импортов

ValadzkoAliaksei commented on src/components/cart/Step3.js:
```diff
@@ -0,0 +1,210 @@
+          </label>
+        </div>
+
+        {(paymentMethod === "visa" || paymentMethod === "masterCard") && (
```
тестовые переменные лучше вынести в константы в виде объекта и брать их значение по ключу объекта

ValadzkoAliaksei commented on src/components/header/Header.js:
```diff
@@ -34,12 +34,12 @@ export default function Header() {
-    ) {
-      setIsShowCart(false);
-    }
+    // if (
```
уберите комментарии

ValadzkoAliaksei commented on src/redux/reducers/orderSlice.js:
```diff
@@ -0,0 +1,54 @@
+  reducers: {
+    // actions
+    reduxGetOrderCountries(state, action) {
+      state.isLoading = "countries";
```
тестовые переменные лучше вынести в константы в виде объекта и брать их значение по ключу объекта

graffad committed:
global-fix-v1

FIXED:
- body scroll is off on showCart
- images import order

CHANGES:
- added constant of types for cart
- added constant for cart default values
- added constant for modalReview default values
- changed cond. classNames with same lib.

REMOVED:
- all unnecessary comments
- console logs

graffad committed:
split-styles
